smart_contract/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+//! API which simplifies writing of smartcontracts
+#![no_std]
+#![allow(unsafe_code)]
+
+extern crate alloc;
+
+use alloc::{boxed::Box, collections::BTreeMap};
+
+#[cfg(not(test))]
+use data_model::smart_contract::payloads;
+use data_model::{
+    isi::Instruction,
+    prelude::*,
+    query::{Query, QueryBox},
+};
+pub use iroha_data_model as data_model;
+pub use iroha_smart_contract_derive::main;
+pub use iroha_smart_contract_utils::{debug, log};
+use iroha_smart_contract_utils::{
+    debug::DebugExpectExt as _, decode_with_length_prefix_from_raw, encode_and_execute,
+};
+use parity_scale_codec::{DecodeAll, Encode};
+
+#[no_mangle]
+extern "C" fn _iroha_smart_contract_alloc(len: usize) -> *const u8 {
+    if len == 0 {
+        iroha_smart_contract_utils::debug::dbg_panic("Cannot allocate 0 bytes");
+    }
+    let layout = core::alloc::Layout::array::<u8>(len).dbg_expect("Cannot allocate layout");
+    // Safety: safe because `layout` is guaranteed to have non-zero size
+    unsafe { alloc::alloc::alloc_zeroed(layout) }
+}
+
+/// # Safety
+/// - `offset` is a pointer to a `[u8; len]` which is allocated in the WASM memory.
+/// - This function can't call destructor of the encoded object.
+#[no_mangle]
+unsafe extern "C" fn _iroha_smart_contract_dealloc(offset: *mut u8, len: usize) {
+    let _box = Box::from_raw(core::slice::from_raw_parts_mut(offset, len));
+}
+
+/// Implementing instructions can be executed on the host
+pub trait ExecuteOnHost: Instruction {
+    /// Execute instruction on the host
+    ///
+    /// # Errors
+    ///
+    /// - If instruction validation failed
+    /// - If instruction execution failed
+    fn execute(&self) -> Result<(), ValidationFail>;
+}
+
+/// Implementing queries can be executed on the host
+pub trait QueryHost: Query {
+    /// Execute query on the host
+    ///
+    /// # Errors
+    ///
+    /// - If query validation failed
+    /// - If query execution failed
+    fn execute(&self) -> Result<Self::Output, ValidationFail>;
+}
+
+// TODO: Remove the Clone bound. It can be done by custom serialization to InstructionExpr
+impl<I: Instruction + Encode + Clone> ExecuteOnHost for I {
+    fn execute(&self) -> Result<(), ValidationFail> {
+        #[cfg(not(test))]
+        use host::execute_instruction as host_execute_instruction;
+        #[cfg(test)]
+        use tests::_iroha_smart_contract_execute_instruction_mock as host_execute_instruction;
+
+        // TODO: Redundant conversion into `InstructionExpr`
+        let isi_box: InstructionExpr = self.clone().into();
+        // Safety: `host_execute_instruction` doesn't take ownership of it's pointer parameter
+        unsafe {
+            decode_with_length_prefix_from_raw(encode_and_execute(
+                &isi_box,
+                host_execute_instruction,
+            ))
+        }
+    }
+}
+
+// TODO: Remove the Clone bound. It can be done by custom serialization/deserialization to QueryBox
```

Comment:
I understand that this and other todos around what used to be the `wasm` crate hail from the expression validator PR, but I didn't see any issues created for it AFAICT. Aren't they somewhat forgotten?

Reply:
opened #3972